File: desktop/packages/mullvad-vpn/src/shared/notifications/forum-activity.ts
import { sprintf } from 'sprintf-js';

import { UNREAD_SATURATED } from '../forum-identity';
import { messages } from '../gettext';
import { RoutePath } from '../routes';
import {
  SystemNotification,
  SystemNotificationCategory,
  SystemNotificationProvider,
  SystemNotificationSeverityType,
} from './notification';

interface ForumActivityNotificationContext {
  unread: number;
}

/**
 * Desktop banner for new community-forum activity.
 *
 * Severity `low` on purpose: that is the band the global notification
 * setting is allowed to suppress, so turning system notifications off
 * silences this too, and the forum-specific setting is an extra gate on
 * top rather than a way around it.
 *
 * The count is all the broadcast digest carries. Nothing here names a
 * topic or an author, which is what keeps the badge free of any per-user
 * request; the content is only ever read when the user opens the panel.
 */
export class ForumActivityNotificationProvider implements SystemNotificationProvider {
  public constructor(private context: ForumActivityNotificationContext) {}

  public mayDisplay() {
    return this.context.unread > 0;
  }

  public getSystemNotification(): SystemNotification {
    return {
      message: this.systemMessage(),
      category: SystemNotificationCategory.forumActivity,
      severity: SystemNotificationSeverityType.low,
      action: {
        type: 'navigate-internal',
        link: {
          // TRANSLATORS: Button on the notification about new forum activity,
          // TRANSLATORS: which opens the activity panel in the app.
          text: messages.pgettext('notifications', 'Open'),
          to: RoutePath.forumActivity,
        },
      },
    };
  }

  private systemMessage(): string {
    const { unread } = this.context;

    // One digest character per slot, so the count stops climbing at its
    // ceiling. Saying "15" there would be a number the user can check and
    // find wrong.
    if (unread >= UNREAD_SATURATED) {
      return sprintf(
        // TRANSLATORS: Notification shown when a lot of forum activity is
        // TRANSLATORS: waiting, above the count the app can measure exactly.
        // TRANSLATORS: Available placeholders:
        // TRANSLATORS: %(count)d - the highest count the app can measure
        messages.pgettext('notifications', 'More than %(count)d new notifications on the forum'),
        { count: unread - 1 },
      );
    }

    return sprintf(
      messages.npgettext(
        'notifications',
        // TRANSLATORS: Notification shown when one reply, like or mention is
        // TRANSLATORS: waiting on the community forum.
        'New notification on the forum',
        // TRANSLATORS: Notification shown when several replies, likes or
        // TRANSLATORS: mentions are waiting on the community forum.
        // TRANSLATORS: Available placeholders:
        // TRANSLATORS: %(count)d - how many are waiting
        '%(count)d new notifications on the forum',
        unread,
      ),
      { count: unread },
    );
  }
}

File: android/lib/model/src/main/kotlin/com/warrenbrowse/vpn/lib/model/forum/ForumActivity.kt
package com.warrenbrowse.vpn.lib.model.forum

/**
 * How the broadcast activity digest becomes one user's badge, and what the
 * header's forum slot carries. Mirrors the desktop `forum-identity.ts`, the
 * one place the three surfaces (bell, notification, launcher badge) take
 * their rules from, so none of them can be left behind the others.
 */

/** Highest count one digest character can carry, above which it saturates. */
const val UNREAD_SATURATED = 15

/**
 * Unread count this installation's [slot] carries in [digest] (one lowercase
 * hex character per slot).
 *
 * Zero whenever there is nothing to show: no fresh document, no slot yet, or a
 * slot past what the server has published (the normal state of an account
 * that registered since the last rebuild). The Rust side has already checked
 * the signature and the freshness, so this only indexes.
 */
fun unreadForSlot(digest: String?, slot: Int?): Int =
    digest?.getOrNull(slot ?: -1)?.digitToIntOrNull(radix = 16) ?: 0

/** The count as shown, saturating rather than growing the badge. */
fun unreadLabel(unread: Int): String =
    if (unread >= UNREAD_SATURATED) "$UNREAD_SATURATED+" else unread.toString()

/**
 * Whether the app shows forum ACTIVITY at all: the header bell, the local
 * notification and the launcher badge alike. Off means off everywhere, and a
 * wallet with no forum account has no activity to report.
 */
fun showsForumActivity(hasAccount: Boolean, enabled: Boolean): Boolean = hasAccount && enabled

/** What the header's forum slot carries: the bell, the lifebuoy, or nothing. */
enum class ForumHeaderButton {
    ACTIVITY,
    COMMUNITY,
    NONE,
}

/**
 * Which button the header's forum slot shows.
 *
 * A wallet with no forum account gets a lifebuoy straight to the forum rather
 * than an empty slot: the bell would be inert for them, but the forum is the
 * one thing they might actually want, and it is where an account comes from.
 * The setting still governs the whole slot, lifebuoy included, so a user who
 * turned the forum off gets no forum in their header either.
 */
fun forumHeaderButton(hasAccount: Boolean, enabled: Boolean): ForumHeaderButton =
    when {
        !enabled -> ForumHeaderButton.NONE
        hasAccount -> ForumHeaderButton.ACTIVITY
        else -> ForumHeaderButton.COMMUNITY
    }

/**
 * What the notification about a rise says. One digest character per slot, so
 * the count stops climbing at its ceiling; saying "15" there would be a number
 * the user can check and find wrong, hence [MoreThan] with the highest count
 * the app can measure exactly.
 */
sealed interface ForumActivityWording {
    data object Single : ForumActivityWording

    data class Several(val count: Int) : ForumActivityWording

    data class MoreThan(val count: Int) : ForumActivityWording
}

/**
 * Whether a notification's age still reads better as "2 h ago" than as a
 * date: past a week, "5 weeks ago" tells a reader less than the day it
 * happened (the desktop `relativeTime` threshold).
 */
fun forumNotificationAgeIsRelative(createdAtSecs: Long, nowSecs: Long): Boolean =
    nowSecs - createdAtSecs < RELATIVE_AGE_LIMIT_SECS

private const val RELATIVE_AGE_LIMIT_SECS = 7L * 24 * 3600

fun forumActivityWording(unread: Int): ForumActivityWording =
    when {
        unread >= UNREAD_SATURATED -> ForumActivityWording.MoreThan(UNREAD_SATURATED - 1)
        unread == 1 -> ForumActivityWording.Single
        else -> ForumActivityWording.Several(unread)
    }
